File: Merge Two Sorted Lists.py
# Link: https://leetcode.com/problems/merge-two-sorted-lists/submissions/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    
    def printList(self, l1):
        
        output = ""
        while l1 != None:
            output += "{} ".format(l1.val)
            l1 = l1.next
            
        print(output)
    
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        
        # Base cases: Exit if one or both list are empty
        if l1 == None and l2 == None:
            return l1
        
        if l1 == None and l2 != None:
            return l2
        
        if l1 != None and l2 == None:
            return l1
        
        # Create new nodes and a copy head node
        currNode = ListNode()
        headNode = currNode
        
        # Compare both lists as long as they're not empty
        while l1 != None and l2 != None:
            
            if l1.val <= l2.val:
                
                # Link currNode to current l1 node
                currNode.next = l1
                
                # Update l1 node
                l1 = l1.next
                
            else:
                
                # Link currNode to current l2 node
                currNode.next = l2
                
                # Update l2 node
                l2 = l2.next
                
            # Move to next node
            currNode = currNode.next
                
        # Add remaining nodes from either list
        if l1 != None:
            currNode.next = l1
        elif l2 != None:
            currNode.next = l2
            
        return headNode.next

File: Merge Two Sorted Lists.java
/*
Link: https://leetcode.com/problems/merge-two-sorted-lists/

Time Complexity: O(n) where n = max(len(list1), len(list2)). 

Space Complexity: O(1). I did not create a new linked list for this problem

Thought process:
- If both lists are null, exit function
- If only 1 list is null, return the non-empty list
- Create a blank node that will link to the head. I'll use it to return the head at the end of the function
- Iterate list 1 and 2
- While iterating, connect to the next smallest node and check that next node in both lists isn't null
- If one list is shorter than the other, add the remaining elements from the long list after adding all elements from shorter list
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null && list2 == null) {
            return list1;
        } else if (list1 == null) {
            return list2;
        } else if (list2 == null) {
            return list1;
        }

        // Create 2 new nodes for future use
        ListNode blankNode = new ListNode();
        ListNode curNode = new ListNode();
        
        // Create an iterate for both linked list
        ListNode itA = list1;
        ListNode itB = list2;

        // Immediately decide what will be the new head and move curNode
        if (itA.val <= itB.val) {
            curNode = itA;
            itA = itA.next;
        } else {
            curNode = itB;
            itB = itB.next;
        }

        // Link blankNode to curNode. blankNode will help return the new head
        blankNode.next = curNode;

        // Iterate list
        while (itA != null && itB != null) {

            // Connect smallest node to curNode and move curNode
            if (itA.val <= itB.val) {
                curNode.next = itA;
                curNode = curNode.next;

                itA = itA.next;
            } else {
                curNode.next = itB;
                curNode = curNode.next;

                itB = itB.next;
            }
        }

        // Add remaining elements
        if (itA != null) {
            curNode.next = itA;
        } else if (itB != null) {
            curNode.next = itB;
        }

        return blankNode.next;
    }
}
